core/src/chain_id.rs

```diff
@@ -0,0 +1,160 @@
+use mina_p2p_messages::v2::{
+    MinaBaseProtocolConstantsCheckedValueStableV1, StateHash, UnsignedExtendedUInt32StableV1,
+};
+use multihash::{Blake2b256, Hasher};
+use time::macros::format_description;
+use time::OffsetDateTime;
+
+use std::fmt::{self, Debug, Display, Formatter};
+
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+#[derive(Clone, PartialEq, Eq)]
+pub struct ChainId([u8; 32]);
+
+fn u8_slice_as_hex(slice: &[u8]) -> String {
+    slice.iter().map(|b| format!("{:02x}", b)).collect()
+}
+
+fn md5_hash(data: u8) -> String {
+    let mut hasher = md5::Context::new();
+    hasher.consume(data.to_string().as_bytes());
+    let hash: Md5 = *hasher.compute();
+    u8_slice_as_hex(&hash)
+}
+
+type Md5 = [u8; 16];
+
+fn hash_genesis_constants(
+    constants: &MinaBaseProtocolConstantsCheckedValueStableV1,
+    tx_pool_max_size: &UnsignedExtendedUInt32StableV1,
+) -> [u8; 32] {
+    let mut hasher = Blake2b256::default();
+    let genesis_timestamp = OffsetDateTime::from_unix_timestamp_nanos(
+        (constants.genesis_state_timestamp.0 .0 .0 * 1000000) as i128,
+    )
+    .unwrap();
+    let time_format =
+        format_description!("[year]-[month]-[day] [hour]:[minute]:[second].[subsecond digits:6]Z");
+    let input = format!(
```

Comment:
@Sventimir Here a new string is allocated, that contains concatenated bytes. Can you do this by just passing each part to the `hasher.update()`?

Reply:
it will, but what reads better? I don't think these allocations here are something to worry about, this code will be called what, once during startup?

Reply:
oops, sorry, thought this was about the time format above, in this case it is plain concatenation, so `hasher.update()` will read as well as this version.

Reply:
It works like this:
```
    hasher.update(constants.k.to_string().as_bytes());
    hasher.update(constants.slots_per_epoch.to_string().as_bytes());
    hasher.update(constants.slots_per_sub_window.to_string().as_bytes());
    hasher.update(constants.delta.to_string().as_bytes());
    hasher.update(tx_pool_max_size.to_string().as_bytes());
    hasher.update(genesis_timestamp.format(&time_format).unwrap().as_bytes());
```

Reply:
Done.